mlos_bench/mlos_bench/services/base_fileshare.py

```diff
@@ -39,12 +39,10 @@ def __init__(self, config: Optional[Dict[str, Any]] = None,
         parent : Service
             Parent service that can provide mixin functions.
         """
+        # IMPORTANT: Save the local methods before invoking the base class constructor
+        local_methods = [self.upload, self.download]
```

Comment:
```suggestion
        self._local_methods = [self.upload, self.download]
```
?

Not sure that this is super useful, but maybe allows for a little bit more introspection and debugging help later.

Reply:
Actually, if we did that, then we could do the following in the base class:

```python
def __init__(self, ...) -> None:
    # immediately check to see if we're instantiating the base class, or it's a subclass calling super().__init__()
    if self.__class__ == Service:
        self._local_methods = []
    else:
        # Child classes must populate self._local_methods prior to calling super().__init__()
        assert self._local_methods
    ... # other init work
    # At the end automatically register the methods for that child instance.
    self.register(self._local_methods)
```

Now we could drop the `self.register()` call from all the child classes since it should be taken care of in the super class, right?

Reply:
Hmm, that might not work if there's mulitple inheritance and bases classes building up.

Reply:
I left an in progress version of this adaptation here for now:

https://github.com/motus/MLOS/compare/sergiym/env/composite_service...bpkroth:MLOS:sergiym/env/composite_service

Reply:
yeah, that's a great idea! now when I think about it, we can even do without `self._local_methods` at all and just pass that list into the superclass constructor. I'll play with it now

Reply:
Actually, here's a full version:
https://github.com/motus/MLOS/pull/7

Reply:
oops, I've just pushed my own update a minute ago :) @bpkroth can you please take a look?

Reply:
LGTM!  Thanks :)